scaaml/stats/online.py

```diff
@@ -0,0 +1,264 @@
+        """Return the result (here the sum). Or None if no summation was done
+        (unknown shape).
+        """
+        # Beware of `if not self._sum` due to zero.
+        if self._sum is None:
+            # Either one or both are None.
+            return None
+
+        assert self._c is not None
+        return self._sum + self._c
+
+    def update(self, value: np.typing.NDArray[np.float64]) -> None:
+        """Update by one element.
+
+        Args:
+
+          value (np.typing.NDArray[np.float64]): Next value to be summed.
+        """
+        # Beware of `if not self._sum` due to zero.
+        if self._sum is None:
+            # First element.
+            self._sum = np.zeros_like(value, dtype=np.float64)
+            self._c = np.zeros_like(value, dtype=np.float64)
+
+        t = self._sum + value
+
+        # Add compensation for the low order bits which would be lost:
+        # - if self._sum is larger then low order bits of value would be lost
+        # - if value is larger then vice-versa
+        self._c += np.where(
+            np.abs(self._sum) >= np.abs(value),  # condition
+            (self._sum - t) + value,  # value when condition is true
+            (value - t) + self._sum,  # value when condition is false
+        )
+
+        # Update the running sum.
+        self._sum = t
+
+
+class Mean:
+    """Compute mean in an online fashion.
+    """
+
+    def __init__(self) -> None:
+        """Initialize the computation.
+        """
+        # Seen elements.
+        self._elements: int = 0
+        self._sum: Sum = Sum()
+
+    def update(self, value: np.typing.NDArray[np.float64]) -> None:
+        """Update by the next value."""
+        self._elements += 1
+        self._sum.update(value)
+
+    @property
+    def result(self) -> Optional[np.typing.NDArray[np.float64]]:
+        """Return the arithmetic mean if some values have been added using the
+        `update` method. Else return None.
+        """
+        if self._elements:
+            sum_result = self._sum.result
+            assert sum_result is not None
+            return sum_result / self._elements
+        return None
+
+
+class VarianceSinglePass:
+    """Welford's algorithm to compute variance. Could be a little slower (CPU)
+    since each iteration contains a division and multiplication.
+    https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+
+    Example use:
+    ```
+    variance = VarianceSinglePass()
+
+    for value in all_values:
+      variance.update(value)
+
+    print((variance.result - all_values.var(axis=0)).abs().max())
+    ```
+    """
+
+    def __init__(self, ddof: int = 0) -> None:
+        """Initialize a sum.
+
+        Args:
+
+          ddof (int): Degrees of freedom. Defaults to 0. For Bessel's
+          correlation set to 1. For more information see `np.var`.
+        """
+        self._ddof: int = ddof
+        self._mean: Sum = Sum()
```

Comment:
`mean` is `Sum`? That's counter-intuitive considering that in the 2-pass version, `mean` is `Mean`

Reply:
Yes it is a mean which we compute online. First update we set it to the value. Then we update it by (value - mean) / n_seen. Meaning this variable is a running mean without having to represent large numbers. I can update it to Mean, but then I need the delta anyway.